pkg/healthz/healthz.go

```diff
@@ -17,25 +17,100 @@ limitations under the License.
 package healthz
 
 import (
+	"bytes"
+	"fmt"
 	"net/http"
+	"sync"
 )
 
+var (
+	// guards names and checks
+	lock = sync.RWMutex{}
+	// used to ensure checks are performed in the order added
+	names  = []string{}
+	checks = map[string]*healthzCheck{}
+)
+
+func init() {
+	http.HandleFunc("/healthz", handleRootHealthz)
+	// add ping health check by default
+	AddHealthzFunc("ping", func(_ *http.Request) error {
+		return nil
+	})
+}
+
+// AddHealthzFunc adds a health check under the url /healhz/{name}
+func AddHealthzFunc(name string, check func(r *http.Request) error) {
+	lock.Lock()
+	defer lock.Unlock()
+	if _, found := checks[name]; !found {
+		names = append(names, name)
+	}
+	checks[name] = &healthzCheck{name, check}
+}
+
+// InstallHandler registers a handler for health checking on the path "/healthz" to mux.
+func InstallHandler(mux mux) {
```

Comment:
I agree this is a bit annoying. There are two reasons I did this. First, healthz needs to work for the default net/http.mux as well as for any abritrary mux (kubernetes uses both). Second, this allows us to register default health checks (right now only ping but later I could see watch health, apiserver connection health on this list as well).

I couldn't think of a good way to support both these cases with one function without complicating the healthz.go code, but I'll can revisit when I have some bandwidth.